File: Assignment4.java

import java.util.*;

public class Assignment4 {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Random random = new Random();

        String[] DAYS = {"Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"};
        String[] SHIFTS = {"Morning", "Afternoon", "Evening"};

        // Data Structure to store schedule
        Map<String, Map<String, List<String>>> schedule = new HashMap<>();
        for (String day : DAYS) {
            schedule.put(day, new HashMap<>());
            for (String shift : SHIFTS) {
                schedule.get(day).put(shift, new ArrayList<>());
            }
        }

        Map<String, Integer> employeeDays = new HashMap<>();

        // Input Employee Data
        System.out.print("Enter number of employees: ");
        int numEmployees = scanner.nextInt();
        scanner.nextLine(); // Consume newline

        Map<String, List<Pair<String, String>>> employees = new HashMap<>();

        for (int i = 0; i < numEmployees; i++) {
            System.out.print("Enter employee name: ");
            String name = scanner.nextLine();
            List<Pair<String, String>> preferences = new ArrayList<>();
            for (String day : DAYS) {
                System.out.printf("Preferred shift for %s (Morning/Afternoon/Evening): ", day);
                String pref = scanner.nextLine();
                preferences.add(new Pair<>(day, pref));
            }
            employees.put(name, preferences);
            employeeDays.put(name, 0); // Initialize shift count to 0
        }

        // Scheduling Logic
        for (String day : DAYS) {
            for (String shift : SHIFTS) {
                // Filter employees who prefer this shift
                List<String> candidates = new ArrayList<>();
                for (Map.Entry<String, List<Pair<String, String>>> entry : employees.entrySet()) {
                    String name = entry.getKey();
                    List<Pair<String, String>> prefs = entry.getValue();
                    if (prefs.contains(new Pair<>(day, shift)) && employeeDays.get(name) < 5) {
                        candidates.add(name);
                    }
                }
                Collections.shuffle(candidates);

                while (schedule.get(day).get(shift).size() < 2) {
                    if (!candidates.isEmpty()) {
                        String employee = candidates.remove(0);
                        if (employeeDays.get(employee) < 5) {
                            schedule.get(day).get(shift).add(employee);
                            employeeDays.put(employee, employeeDays.get(employee) + 1);
                        }
                    } else {
                        // Randomly assign if fewer than 2 employees
                        List<String> available = new ArrayList<>();
                        for (Map.Entry<String, Integer> entry : employeeDays.entrySet()) {
                            if (entry.getValue() < 5) {
                                available.add(entry.getKey());
                            }
                        }
                        if (!available.isEmpty()) {
                            String employee = available.get(random.nextInt(available.size()));
                            schedule.get(day).get(shift).add(employee);
                            employeeDays.put(employee, employeeDays.get(employee) + 1);
                        } else {
                            break;
                        }
                    }
                }
            }
        }

        // Display Final Schedule
        System.out.println("\nFinal Schedule:");
        for (String day : DAYS) {
            System.out.println("\n" + day + ":");
            for (String shift : SHIFTS) {
                List<String> names = schedule.get(day).get(shift);
                String result = names.isEmpty() ? "No employees assigned" : String.join(", ", names);
                System.out.printf("  %s: %s%n", shift, result);
            }
        }

        scanner.close();
    }

    // Helper class to represent a pair of day and shift
    static class Pair<T, U> {
        T first;
        U second;

        public Pair(T first, U second) {
            this.first = first;
            this.second = second;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            Pair<?, ?> pair = (Pair<?, ?>) o;
            return Objects.equals(first, pair.first) &&
                    Objects.equals(second, pair.second);
        }

        @Override
        public int hashCode() {
            return Objects.hash(first, second);
        }
    }
}

File: assignment4.py
# Employee Scheduler in Python

import random
from collections import defaultdict

# Constants
DAYS = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
SHIFTS = ['Morning', 'Afternoon', 'Evening']

# Data Structure to store schedule
schedule = {day: {shift: [] for shift in SHIFTS} for day in DAYS}
employee_days = defaultdict(int)

# Input Employee Data
num_employees = int(input('Enter number of employees: '))
employees = {}

for _ in range(num_employees):
    name = input('Enter employee name: ')
    preferences = []
    for day in DAYS:
        pref = input(f'Preferred shift for {day} (Morning/Afternoon/Evening): ')
        preferences.append((day, pref))
    employees[name] = preferences

# Scheduling Logic
for day in DAYS:
    for shift in SHIFTS:
        # Filter employees who prefer this shift
        candidates = [name for name, prefs in employees.items() if (day, shift) in prefs and employee_days[name] < 5]
        random.shuffle(candidates)

        while len(schedule[day][shift]) < 2:
            if candidates:
                employee = candidates.pop()
                if employee_days[employee] < 5:
                    schedule[day][shift].append(employee)
                    employee_days[employee] += 1
            else:
                # Randomly assign if fewer than 2 employees
                available = [name for name in employees if employee_days[name] < 5]
                if available:
                    employee = random.choice(available)
                    schedule[day][shift].append(employee)
                    employee_days[employee] += 1
                else:
                    break

# Display Final Schedule
print('\nFinal Schedule:')
for day in DAYS:
    print(f'\n{day}:')
    for shift, names in schedule[day].items():
        print(f'  {shift}: {', '.join(names) if names else 'No employees assigned'}')
